JMMServer/Tasks/RecreateAllGroupsTask.cs

```diff
@@ -0,0 +1,373 @@
+            // The reason we're doing this in parallel is because updating contacts does a reasonable amount of work (including LZ4 compression)
+            Parallel.ForEach(series.Batch(DefaultBatchSize), new ParallelOptions { MaxDegreeOfParallelism = 4 },
+                localInit: () => DatabaseFactory.SessionFactory.OpenStatelessSession(),
+                body: (seriesBatch, state, localSession) =>
+                    {
+                        AnimeSeries.BatchUpdateContracts(localSession.Wrap(), seriesBatch);
+                        return localSession;
+                    },
+                localFinally: localSession => { localSession.Dispose(); });
+
+            _animeSeriesRepo.UpdateBatch(session, series);
+            _log.Info("AnimeSeries contracts have been updated");
+        }
+
+        private void UpdateAnimeGroupsAndTheirContracts(ISessionWrapper session, IReadOnlyCollection<AnimeGroup> groups)
+        {
+            _log.Info("Updating statistics and contracts for AnimeGroups");
+
+            var allCreatedGroupUsers = new ConcurrentBag<List<AnimeGroup_User>>();
+
+            // Update batches of AnimeGroup contracts in parallel. Each parallel branch requires it's own session since NHibernate sessions aren't thread safe.
+            // The reason we're doing this in parallel is because updating contacts does a reasonable amount of work (including LZ4 compression)
+            Parallel.ForEach(groups.Batch(DefaultBatchSize), new ParallelOptions { MaxDegreeOfParallelism = 4 },
+                localInit: () => DatabaseFactory.SessionFactory.OpenStatelessSession(),
+                body: (groupBatch, state, localSession) =>
+                    {
+                        var createdGroupUsers = new List<AnimeGroup_User>(groupBatch.Length);
+
+                        // We shouldn't need to keep track of updates to AnimeGroup_Users in the below call, because they should have all been deleted,
+                        // therefore they should all be new
+                        AnimeGroup.BatchUpdateStats(groupBatch, watchedStats: true, missingEpsStats: true, createdGroupUsers: createdGroupUsers);
+                        allCreatedGroupUsers.Add(createdGroupUsers);
+                        AnimeGroup.BatchUpdateContracts(localSession.Wrap(), groupBatch, updateStats: true);
+
+                        return localSession;
+                    },
+                localFinally: localSession => { localSession.Dispose(); });
+
+            _animeGroupRepo.UpdateBatch(session, groups);
+            _log.Info("AnimeGroup statistics and contracts have been updated");
+
+            _log.Info("Creating AnimeGroup_Users and updating plex/kodi contracts");
+
+            // The reason we're doing this in parallel is because updating contacts does a reasonable amount of work (including LZ4 compression)
+            Parallel.ForEach(allCreatedGroupUsers.SelectMany(groupUsers => groupUsers), groupUser =>

```

Comment:
This and the below Parallel didn't use NHibernate sessions (like the above two) as far as I could tell by tracing through the code, so didn't see any reason to impose a limit.